File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    void dfs(int[] candidates, int idx, ArrayList<Integer> cur, int target, List<List<Integer>> sol){
        if(idx == candidates.length){
            if(target==0)
                sol.add(new ArrayList<Integer>(cur));
            return;
        }
        if(candidates[idx]<=target){
            cur.add(candidates[idx]);
            dfs(candidates, idx, cur, target-candidates[idx], sol);
            cur.remove(cur.size()-1);
        }
        dfs(candidates, idx+1, cur, target, sol);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> sol = new ArrayList<>();
        dfs(candidates, 0, new ArrayList<Integer>(), target, sol);
        return sol;
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void dfs(int idx, vector<int>& cur, vector<int>& candidates, vector<vector<int>>& sol, int target){
        if(target == 0){
            sol.push_back(cur);
            return;
        }
        else if(target < 0)
            return;
        
        if(idx==candidates.size())
            return;
        
        cur.push_back(candidates[idx]);
        dfs(idx, cur, candidates, sol, target-candidates[idx]);
        cur.pop_back();
        dfs(idx+1, cur, candidates, sol, target);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> cur;
        vector<vector<int>> sol;
        dfs(0, cur, candidates, sol, target);
        return sol;
    }
};